### ISSUE TITLE
Remember the original unicode escape sequence

### ISSUE THREAD

**Issue by ftomassetti** [javaparser/javaparser#801]
Do we want to store the very original code, including the unicode escape sequences?
@cal101 provided a patch for that

https://github.com/ftomassetti/javaparser/pull/5/files

@matozoid is in a better position than me to comment on it

**Comment by matozoid** [comment]
Yes we do, and credits to @cal101 for figuring it out, but it does need some work to go from POC to PR. With patches and scripts it's going to be a pain to maintain.

**Comment by cal101** [comment]
I thought it was obvious from my "PR" request comment that I don't expect the patch variant to be an acceptable solution.
I guess that it should be possible to let jvacc use our own version of javacharparser class because thats the way most other parser generators i used before allow it.
I don't know javacc and didn't looked into it because I thought someone from the core team would know that out-of-the-box. If I am wrong I can take a look at it.
The other major part that bothers me is spending cpu cycles and memory for a 5% or less use case in 100% of all cases. For someone knowing the architecture beter than me it should be a no brainer to pass the necessary information from lexical preserving code to the parser to switch unicode preserving on/off as needed.

**Comment by cal101** [comment]
Another point worth mentioning is the textelementxxx class. Does it need only the original code or also the preprocessed one?

**Comment by matozoid** [comment]
I'm looking at this now. It's cool to finally have a fix for something we considered unfixable :)

If I understand correctly the main part of the solution is a custom stream wrapping the usual stream that picks up all the unicode characters and stores them in a new originalText field in JavaToken?

Also: I don't see the build-it script anymore, so some work was done on that? It looks fine to me. I think that answers where we can set the CharParser (in the generated parser's constructor.)

For architecture and optimization I would say: the default for JP is to have all bells and whistles turned on, and you can turn them off with flags in `ParserConfiguration` - so the answer could simply be "add a flag in ParserConfiguration and only wrap the stream when that flag is set.

Textelement seems to be a lexical preservation thing - @ftomassetti knows about that.

One request: there are some remarks about things that do not work 100% (the one with the >>> comes to mind) - if you can add those to the issue tracker then we won't forget about them. You can do that with the on/off flag too if you like.

Oh, one more request: can you make a PR directly to the main project? If not, no problem, I'll do it for you.

Thank you very much!

**Comment by cal101** [comment]
I replaced the patching stuff with direct javacc parser changes. The parser generation config was changed to not generate a JavaCharStream instance but use the simpler CharParser instead.
I modified the former generated JavaCharStream (?) and just renamed. There is no additional wrapper involved.
The changes are small. For review I suggest comparing with the master generated JavaCharStream for the changes I did to that code. Not that many.
I don't plan to hack on that stuff in the next time.
